system/jhtree/jhtree.cpp

```diff
@@ -1088,45 +1088,51 @@ IKeyIndex *CKeyStore::doload(const char *fileName, unsigned crc, IReplicatedFile
     fname.append(fileName).append('/').append(crc).append('/').append(blockedIOSize);
 
     // MORE - holds onto the mutex way too long
-    synchronized block(mutex);
-    keyIndex = keyIndexCache.query(fname);
-    if (NULL == keyIndex)
+    bool isNew = false;
     {
-        if (iMappedFile)
-        {
-            assert(!iFileIO && !part);
-            keyIndex = new CMemKeyIndex(getUniqId(fileIdx), LINK(iMappedFile), fname, isTLK);
-        }
-        else if (iFileIO)
-        {
-            assert(!part);
-            keyIndex = new CDiskKeyIndex(getUniqId(fileIdx), LINK(iFileIO), fname, isTLK, blockedIOSize);
-        }
-        else
+        synchronized block(mutex);
+        keyIndex = keyIndexCache.query(fname);
+        if (NULL == keyIndex)
         {
-            assert(fileIdx==(unsigned) -1);
-            Owned<IFile> iFile;
-            if (part)
+            isNew = true;
+            if (iMappedFile)
             {
-                iFile.setown(part->open());
-                if (NULL == iFile.get())
-                    throw MakeStringException(0, "Failed to open index file %s", fileName);
+                assert(!iFileIO && !part);
+                keyIndex = new CMemKeyIndex(getUniqId(fileIdx), LINK(iMappedFile), fname, isTLK);
+            }
+            else if (iFileIO)
+            {
+                assert(!part);
+                keyIndex = new CDiskKeyIndex(getUniqId(fileIdx), LINK(iFileIO), fname, isTLK, blockedIOSize);
             }
             else
-                iFile.setown(createIFile(fileName));
-            IFileIO *fio = iFile->open(IFOread);
-            if (fio)
-                keyIndex = new CDiskKeyIndex(getUniqId(fileIdx), fio, fname, isTLK, blockedIOSize);
-            else
-                throw MakeStringException(0, "Failed to open index file %s", fileName);
+            {
+                assert(fileIdx==(unsigned) -1);
+                Owned<IFile> iFile;
+                if (part)
+                {
+                    iFile.setown(part->open());
+                    if (NULL == iFile.get())
+                        throw MakeStringException(0, "Failed to open index file %s", fileName);
+                }
+                else
+                    iFile.setown(createIFile(fileName));
+                IFileIO *fio = iFile->open(IFOread);
+                if (fio)
+                    keyIndex = new CDiskKeyIndex(getUniqId(fileIdx), fio, fname, isTLK, blockedIOSize);
+                else
+                    throw MakeStringException(0, "Failed to open index file %s", fileName);
+            }
+            keyIndexCache.replace(fname, *LINK(keyIndex));
+        }
+        else
+        {
+            LINK(keyIndex);
         }
-        keyIndexCache.replace(fname, *LINK(keyIndex));
-    }
-    else
-    {
-        LINK(keyIndex);
     }
     assertex(NULL != keyIndex);
+    if (isNew && recordingEvents())
+        queryRecorder().recordFileInformation(keyIndex->queryId(), fileName);
```

Comment:
Just an observation... By recording meta info after events using it, event visitors will be more complicated because cross references can't be assumed to be available. There will need to be event accumulation, in some form, until cross references can be resolved.

Reply:
Yes, agreed.  I think on balance it is better to enforce the ordering.